lib/utility.js

```diff
@@ -1,10 +1,171 @@
+      config['data'] = data;
+    }else{
+      config['filePath'] = filePath;
+      config['formData'] = data;
+      config['name'] = "ProductImage";
+    }
+    req(config);
+  });
+}
 
 /**
+ * Choose images (Max: 9)
+ * @promise resolve: the tmp imgPaths (Array[String])
+ * @promise reject: Fail to extract images
+ */
+function chooseImagePromise(){
+  let config = {
+    "count": 9, // at-most 9 images at onces selected
+    "sizeType": "original",
+    "sourceType": "album"
+  }
+  return new Promise(function(resolve, reject){
+    config.success = res => {resolve(res.tempFilePaths)};
+    config.fail = err => {reject(err)};
+    wx.chooseImage(config);
+  })
+}
+
+/** @review
+ * Promisified function for user login
+ * Also can be used to check existence of globalData.userInfo
+ * @param config (Optional): the globalData object used to automatically store the retrieved UserID 
+ * @promise resolve: UserId (unique) from server
+ * @promise reject:  Fail to generate wx code or fail to retrieve UserId from server
+ */
+function getUserId(globalData){ 
+  return new Promise(function(resolve, reject){
+    var app = getApp();
+    // In case userId already exists
+    if (app && app.globalData.userInfo) return resolve(app.globalData.userInfo);
+    wx.login({
+      success: function(res){
+        if (!res.code) reject({"ErrorMessage": "Fail to Generate Login Code"});
+        requestPromise(loginUrl, {"LoginCode": res.code})
+        .then(function(data){
+          app = getApp();
+          if (globalData) globalData.userInfo = data.UserId;
+          if (app) app.globalData.userInfo = data.UserId;
+          return resolve(data.userId);
+         })
+      },
+      fail: function(res){
+        reject({'message': "Fail to Login"});
+      }
+    });
+  })
+}
+
+
+/** @review
+ * Query Server for my product list 
+ * @param offset: request offset
+ * @param offset: request pageSize
+ */
+function getMyProduct(offset, pageSize){
+  return getUserId(getApp().globalData) // Make sure UserId is retrieved
+          .then((id)=>{
+            let data = {
+              "Action": "GetProductsByUserId",
+              "UserId": id
+            };
+            if (offset) data["StartId"] = offset;
+            if (pageSize) data["pageSize"] = pageSize;
+            return requestPromise(marketUrl, data);
+          })
+          .then((res)=>{ return (res.Products.map((x) => new Product(x))); });
+}
+
+/** @review
+ * post 1 product and its images (auxiliary)
+ * @param data.ProductName: String 
+ * @param data.ProductType: Number
+ * @param data.ProductImages: Array[String] (OPTIONAL)
+ * @promise resolve: Server response
+ * @promise reject: Lack of Info
+ */
+function postProduct(data) {
+  // Check validity
+  if (!data.ProductName)
+      throw { 'message': "Lack product name" };
+  if (!data.ProductType)
+      throw { 'message': "Lack product type" };
+  data.ProductOwner = getApp().globalData.userInfo;

```

Comment:
I think here we should also use getUserId() like what we did in getMyProduct()

It is possible user id has not been fetched when this function is being called.